BIP375: Sending Silent Payments in PSBTs

andrewtoth opened the pull request:
This BIP adds support for sending silent payments using PSBTs.

If there are multiple entities handling the PSBT that do not have access to some input private keys, a DLEQ proof by the signer may be added for other entities to verify the corresponding ECDH shares used to derive the output scripts were generated correctly. This is specified in BIP374. For the common case of a single entity that has access to all private keys, the DLEQ proof generation is unnecessary.

Spending support is trivial and can be done with a modification to BIP370 to add a new input field for the tweak data.

jonatack commented:
Mailing list post on Oct 17 at https://groups.google.com/g/bitcoindev/c/5G5wzqUXyk4.

jonatack commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+| 0
+| 2
+|-
+| Silent Payment Global DLEQ Proof
```
Suggest defining "Discrete Log Equality Proofs (DLEQ)" above this / on first use, and discuss overlap with #1689 (if any) or linkage to be made between the two.

andrewtoth replied:
I link to BIP374 and added it as a dependency as well.

murchandamus commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+
+This document specifies new fields and new field inclusion/exclusion requirements.
+
+<tt>PSBT_OUT_SCRIPT</tt> is modified to be optional for outputs in silent payments capable PSBTs. If this field is not included in the output, then the field PSBT_OUT_SP_V0_INFO must be included.
```
Does it make sense to require the presence of either `PSBT_OUT_SCRIPT` or `PSBT_OUT_SP_V0_INFO`? What if there is another future extension of BIP 370 that provides a third option?

Would it perhaps be sufficient to declare an output that has both the fields `PSBT_OUT_SCRIPT` and `PSBT_OUT_SP_V0_INFO` to make the PSBT invalid (i.e. at most one of these two can be present)?

Perhaps the interplay of these two fields could be elaborated.

andrewtoth replied:
I don't think making these fields exclusive will work. The `PSBT_OUT_SCRIPT` must be added to sign the PSBT. The `PSBT_OUT_SP_V0_INFO` must be present to verify that the `PSBT_OUT_SCRIPT` is actually correct after it is added.

> What if there is another future extension of BIP 370 that provides a third option?

That would indeed soften this requirement. This BIP would have to be modified to support that extension, making it

> If either `PSBT_OUT_SCRIPT` or `PSBT_OUT_NEW_EXTENSION` are not included in the output, then the field PSBT_OUT_SP_V0_INFO must be included.

I don't really have an idea on how this could be made more graceful for that scenario.

> Perhaps the interplay of these two fields could be elaborated.

I will add some more context in a footnote, thanks.

andrewtoth replied:
Clarified this section and added a footnote.

murchandamus commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+
+===Unique Identification===
+
+Silent payment capable PSBTs can be uniquely identified the same way as PSBTv2s, except when including silent payment outputs. For silent payment capable PSBTs, all silent payment outputs must use the PSBT_OUT_SP_V0_INFO instead of PSBT_OUT_SCRIPT as the output script when creating the unsigned transaction used for unique identification.
```
I found this sentence confusing. Does this also hold if both `PSBT_OUT_SP_V0_INFO` and `PSBT_OUT_SCRIPT` are specified on an output?

andrewtoth replied:
No, it does not hold once `PSBT_OUT_SCRIPT` is added to an output. Only outputs without that field would need to fall back to `PSBT_OUT_SP_V0_INFO`. I will clarify this in the text.

andrewtoth replied:
I was incorrect. This *does* also hold if both are specified. I updated this section and added a footnote for the reasoning.

murchandamus commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+===Constructor===
+
+All rules must be followed from PSBTv2 for this role, with the following exception:
+When an output is added, it must have either PSBT_OUT_SCRIPT or PSBT_OUT_SP_V0_INFO, or both, set.
```
I only understood here that both of these fields can be set at the same time. Perhaps that could be explained above where `PSBT_OUT_SP_V0_INFO` is first introduced.

andrewtoth replied:
Added more context above.

murchandamus commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+
+Additionally to PSBTv2, the Constructor must also follow additional rules:
+
+Inputs spending an output with script using Segwit version > 1 may only be added if there are no outputs with PSBT_OUT_SP_V0_INFO set.
+Outputs with PSBT_OUT_SP_V0_INFO set may only be added if there are no inputs spending an output script using Segwit version > 1.
```
This might be a bit out of scope for your PR, but I was just wondering:
Silent Payment transactions disallow inputs spending native segwit outputs with version > 1, but should it perhaps also exclude native segwit outputs with version 1 that are not P2TR inputs? Most of version 1 is yet unencumbered, are Silent Payments robust in regard to Pay to Anchor?

andrewtoth replied:
I believe BIP 352 specifies that only the script template `OP_1 <32 bytes>` is used, and not that it is segwit v1. So, P2A does not follow that template, and will be ignored. But, segwit versions higher than v1 specifically are prohibited for forward compatibility reasons.

This is what I understood from @josibake's explanation.

murchandamus commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+
+The Signer should additionally compute the silent payment addresses, optionally showing this data to the user instead of the computed segwit v1 addresses.
+
+If a sighash type is provided and there are silent payment outputs present, the signer must fail if the sighash type is not SIGHASH_ALL.
```
As mentioned above, this seems to be stricter than BIP 352 itself. Is this intentional?

andrewtoth replied:
This is intentional. `SIGHASH_SINGLE` is unsafe for `k` values. `SIGHASH_NONE` does not make sense since you will have to compute ECDH shares for the outputs, but you presumably don't care about the outputs if you are signing with that flag.

andrewtoth replied:
Added a footnote to clarify this.

murchandamus commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+* Let ''a<sub>n</sub>'' be the sum of the private keys ''a'' of all eligible inputs
+* Let ''C =  a<sub>n</sub>·B<sub>scan</sub>''
+
+Use a key ''B<sub>scan</sub>'' followed by a list of the outpoints of all eligible inputs.
```
I was surprised that we provide a single share and a single DLEQ proof per participant rather than one per input. In case there are multiple participants, this leaks to all other participants which inputs were provided in bulk by one party. I was wondering if there might be better privacy properties if participants provide a separate share for each input, that way all subsequent participants do not know how previous inputs group if the PSBT is passed in a circle rather than shared with all participants after each step. Maybe I’m overthinking this, though.

andrewtoth replied:
This is provided as a performance optimization so that there is only one proof for multiple inputs.
It is not required to combine all your inputs. You could selectively add different shares and proofs for each inputs you don't want to link. I suppose I could make that option more explicit in the text.

andrewtoth replied:
See my comment [here](https://github.com/bitcoin/bips/pull/1687/files#r1896266653).

murchandamus commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+
+====Change Detection====
+
+Updaters may add two PSBT_OUT_BIP32_DERIVATION key-value-pairs with the corresponding derivation path of both the scan and spend keys. The Signer can then use these fields to verify that the silent payment code is change.
```
Should the "Updater" role perhaps appear before the "Signer"? Adding change would probably happen at the same time participants add inputs, so surely before signing?

andrewtoth replied:
Done.

murchandamus commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+  Status: Draft
+  Type: Standards Track
+  Created: 2024-05-14
+  License: BSD-2-Clause
```
```suggestion
  License: BSD-2-Clause
  Post-History: https://groups.google.com/g/bitcoindev/c/5G5wzqUXyk4
  Requires: 352, 370
```

andrewtoth replied:
Done.

murchandamus reviewed (commented):
I did a light review and left a few questions mostly. Your idea makes sense, the content is coming along nicely, and I have found no issues regarding the formatting.

Perhaps some of the answers to my questions could be recorded as footnotes in the Rationale, where they seem likely to be of interest to future readers.

andrewtoth commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+===Motivation===
+
+Partially Signed Bitcoin Transaction Version 2 as described in BIP 370 is not compatible with sending to silent payments as described in BIP352. In particular, the output script of a silent payment cannot be computed until after all transaction inputs have been added.
+Also, any inputs that the Signer has the private keys for must be signed with SIGHASH_ALL and all inputs must not have any scriptPubKeys with Segwit version > 1.
```
Signing with `SIGHASH_NONE` doesn't make sense because you will have to generate the ECDH shares for all payments, but they could then be removed or new ones added. So you will have to go back the signer again. I suppose technically this could be allowed but I don't see a use for it, and it could be used by a griefer to force the party to keep going back to the signer.
Signing with `SIGHASH_SINGLE` would allow another participant to add a payment to the same scan key, but a lower lexicographic order, invalidating the `k` value in the output script.
Signing with `SIGHASH_ANYONECANPAY` is unsafe as per BIP 352.

I suppose I could add this as a footnote.
But also, I don't think this sentence should be in the motivation section.

andrewtoth replied:
Removed the sentence, but also added a footnote in the Signer section to clarify why.

andrewtoth commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+
+Partially Signed Bitcoin Transaction Version 2 as described in BIP 370 is not compatible with sending to silent payments as described in BIP352. In particular, the output script of a silent payment cannot be computed until after all transaction inputs have been added.
+Also, any inputs that the Signer has the private keys for must be signed with SIGHASH_ALL and all inputs must not have any scriptPubKeys with Segwit version > 1.
+Additionally, the silent payment outputs computed by a signer must be verifiable to other entities.
```
> If there is a single signer, why would others need to be able to verify?

In the case of a hardware wallet connected to a software wallet, the hardware wallet is the single signer but the software wallet must verify that the output is computed properly before broadcasting.

> If there are multiple signers, wouldn’t all signers need to collaborate by putting forth shares rather than "computing silent payment outputs"?

Yes, all signers need to collaborate and put forth shares, but they must also compute the output script before signing. This computed output script must be added to the PSBT before signing to be compatible with BIP 370 signing process. After it is added, the other signers can compute the output script to verify themselves before signing.

andrewtoth commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+|-
+| Silent Payment Global ECDH Share
+| <tt>PSBT_GLOBAL_SP_ECDH_SHARE = 0x07</tt>
+| <tt><33 byte scan key> <36 byte outpoint>*</tt>
```
> The main advantage over allowing only one outpoint per ECDH share seems to be a reduction in proving complexity for signers who choose to provide DLEQ proofs

The main advantage is reduction in both ECDH share generation and proving complexity.
All outpoint private keys can be summed, and then only a single ECC mult for the shares, and 2 ECC mults for the proof.

Consider the common case of a single sig wallet using a hardware signing device receiving many small inputs over time. The wallet now contains 100 utxos and wants to spend to a single silent payment address.
By using single share and proof for each outpoint, that is 100 * ECDH share + 100 * DLEQ proof (2 ECC mults) + 100 * signatures = 400 ECC mults.
By summing all inputs, the result is now 1 * ECDH share + 1 * DLEQ proof (2 ECC mults) + 100 signatures = 103 ECC mults.
On a simple hardware signing device, this makes the signing time 4x, making it potentially a few minutes to over 10.

re: disadvantages
1. It is only an option to combine the inputs. You can still provide a share and proof per input or whatever combination of inputs you'd like. I will make this more clear in the text. I believe the single sig with hardware wallet is the most common case though, so we should optimize for that case.
2. This is indeed a potential pitfall. It could also be addressed by providing a share and proof per input as a fallback. I will add some text to address this situation.

andrewtoth replied:
[Here's a blog post](https://blog.casa.io/bitcoin-multisig-hardware-signing-performance-2024/) of time it takes to sign transactions on various hardware devices. I think it makes sense to think about optimizing for this.

andrewtoth replied:
I think the 2 main use cases will be for a wallet that has access to all inputs, or only a few that they would not like to link. So, perhaps having subsets of different inputs is not the best design for this spec.

How about optional fields per input or optional global fields? If the global fields are present, they represent the sum of all inputs. Otherwise, a field per input must be set. I think this is simpler than using the key to specify subsets.

andrewtoth commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+| Silent Payment Global DLEQ Proof
+| <tt>PSBT_GLOBAL_SP_DLEQ = 0x08</tt>
+| <tt><33 byte scan key> <36 byte outpoint>*</tt>
+| The scan key and a list of outpoints corresponding to the prevouts of the inputs that this proof covers. The outpoints are composed of a 32 byte txid followed by a 32-bit little endian uint.
+| <tt><64-byte proof></tt>
+| A DLEQ proof computed for the matching ECDH share.
```
Yes, this is a great insight, thank you!

Would it not also reduce the complexity, since it would only be one proof to verify after summing the `B_scan` generators instead of verifying each proof individually?

andrewtoth replied:
I'm not sure the added complexity is worth it to add this.

andrewtoth commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+ - Verify the DLEQ proofs for all inputs it does not have the private keys for.
+ - If all eligible inputs have an ECDH share, compute and set the PSBT_OUT_SCRIPT.
+
+If the Signer sets any missing PSBT_OUT_SCRIPTs, it must set the Inputs Modifiable flag to False.
+
+If any output does not have PSBT_OUT_SCRIPT set, the Signer must not yet add a signature.
```
I don't think it's this requirement, but the one a few lines up:
`* If all eligible inputs have an ECDH share, compute and set the PSBT_OUT_SCRIPT.`

We would need to only compute the output scripts for the non-blinded outputs.
Yes, if we want to hide which participants add an SP vs a regular taproot address entirely, we would need all outputs to have a dummy SP_V0_INFO and have everyone compute shares and proofs for them, even if they don't end up being used.

andrewtoth commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+===Abstract===
+
+This document proposes additional fields and updated role responsibilities for BIP 370 PSBTv2
+which adds support for sending to silent payments as described in BIP352.
```
Done.

andrewtoth commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,187 @@
+| <tt><33 byte scan key> <36 byte outpoint>*</tt>
+| The scan key and a list of outpoints corresponding to the prevouts of the inputs that this ECDH share is for. The outpoints are composed of a 32 byte txid followed by a 32-bit little endian uint.
+| <tt><32 byte share></tt>
+| An ECDH share for a scan key, followed by a list of outpoints. The ECDH shared is computed with ''a * B_scan'', where ''a'' is the sum of all private keys of the inputs matching the list of outpoints, and ''B_scan'' is the scan key of a recipient.
```
Done.

guggero commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,196 @@
+! Versions Allowing Inclusion
+|-
+| Silent Payment Data
+| <tt>PSBT_OUT_SP_V0_INFO = 0x08</tt>
```
Just a drive-by comment that this value is now taken by `PSBT_OUT_MUSIG2_PARTICIPANT_PUBKEYS` which was added by https://github.com/bitcoin/bips/pull/1540.

guggero commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,196 @@
+| <tt><33 byte scan key> <36 byte outpoint>*</tt>
+| The scan key and a list of outpoints corresponding to the prevouts of the inputs that this ECDH share is for. The outpoints are composed of a 32 byte txid followed by a 32-bit little endian uint.
+| <tt><33 byte share></tt>
+| An ECDH share for a scan key, followed by a list of outpoints. The ECDH shared is computed with ''a * B_scan'', where ''a'' is the sum of all private keys of the inputs matching the list of outpoints, and ''B_scan'' is the scan key of a recipient.
```
I have an implementation question about this. If all the shares are `a * B_scan`, then how can the last signer (or the transaction extractor) calculate the output key? BIP-0352 mentions `input_hash = hash_BIP0352/Inputs(outpointL || A)`, but how would I extract `A` from the shares? Is it as simple as subtracting `B_scan` from the share?

guggero replied:
Ah, I guess a PSBT is supposed to contain all BIP-32 derivation information for all inputs. So `A` can be derived from the sum of all public keys. Is that correct? Then perhaps that could be mentioned here.

andrewtoth replied:
> how would I extract A from the shares?

`A` in the context of the `input_hash` is `a*G`, or the sum of all public keys. This might be easier to understand from the receiver side https://github.com/bitcoin/bips/blob/master/bip-0352.mediawiki#scanning.

andrewtoth replied:
> a PSBT is supposed to contain all BIP-32 derivation information for all inputs

I'm not sure I follow why we need BIP-32 derivation information? The public keys can be extracted by the prevout scripts of the inputs. https://github.com/bitcoin/bips/blob/master/bip-0352.mediawiki#inputs-for-shared-secret-derivation

andrewtoth replied:
I think I understand, for output generation:

`Let ecdh_shared_secret = input_hash·a·Bscan`

replace `a·Bscan` with the ECDH share.

This is detailed [here](https://github.com/bitcoin/bips/pull/1687/files#diff-b41d8882b5f3bcf8086d19f64655573bdb4666f5091054c4b9509b60e185293aR174).

guggero replied:
True, the `PSBT_IN_HASH160` should probably work as well. I just haven't seen this field being set "in the wild" up to now, so not sure if all wallets will set it correctly (at least `lnd` doesn't, which perhaps means it's not fully BIP-0174 compliant?).
So I thought the public key in the BIP-32 derivation fields is probably more likely to be set in most wallets.
But perhaps this BIP doesn't have to go into details around that and can just mention something along the lines of:
"Signers creating the output keys will not be able to extract public keys from the witness as at that point the inputs aren't signed yet. The meta information present in the PSBT should include the public key of each input though, so creating the sum of all input public keys (`A`) should happen based on that information".

andrewtoth replied:
> But perhaps this BIP doesn't have to go into details around that

Hmm I'm not sure that's a good idea to leave that unspecified. Different implementations could then depend on different metadata fields that may or may not be set and be incompatible. The purpose of a standard is to avoid that problem.

If it's easier we can use the BIP-32 derivation field since I see `PSBT_IN_HASH160` is not implemented in https://github.com/bitcoinjs/bip174 either. The difference is that the value of that field should be set to a fingerprint and derivation path which some signers might not want to share. We can specify in this document to set this field for non-taproot sp eligible inputs, and a zero fingerprint and empty path array for the value.

guggero reviewed (commented):
One question about calculating the output keys, other than that this looks good.

andrewtoth commented:
@murchandamus @jonatack @nothingmuch @guggero @theStack thank you all for your reviews.

- Updated the spec to have either per-input or global ECDH shares and DLEQ proofs, instead of using subsets.
- Specified using BIP374 to compute and verify DLEQ proofs.
- Suggest adding BIP32 derivation paths for pubkey hashed inputs, so the public key is available before the witness or scriptSig is added.
- Updated output key due to conflict with #1540.
- Added silent payment label output type for computing change.

jonatack commented on bip-PSBT-SP.mediawiki:
```diff
@@ -0,0 +1,262 @@
+<pre>
+  BIP: ?
```
Assigned 375.

Please update here, add a README entry, and update the `Created` field to today's date.

jonatack reviewed (commented):
Assigned 375.

SupavineeSerksiri6315 commented:
ฉันขอขอบคุณมาก ช่วยแนะนำฉันด้วยค่ะ

jonatack commented on bip-0375.mediawiki:
```diff
@@ -0,0 +1,261 @@
+
+If the Signer has the private keys for all eligible inputs, the Signer should generate a global ECDH share for each scan key ''B<sub>scan</sub>'' as follows:
+
+Using the notation from [https://github.com/bitcoin/bips/blob/master/bip-0352.mediawiki#specification BIP352]
```
nano-nit, the other "Using the natation from BIP352" sentence ends with a colon

```suggestion
Using the notation from [https://github.com/bitcoin/bips/blob/master/bip-0352.mediawiki#specification BIP352]:
```

Sjors commented on bip-0375.mediawiki:
```diff
@@ -0,0 +1,261 @@
+<pre>
+  BIP: 375
+  Layer: Applications
+  Title: Sending Silent Payments with PSBTs
```
I wouldn't rename the title, but instead in the introduction clarify that the minimum PSBT version is 2.

Sjors commented on bip-0375.mediawiki:
```diff
@@ -0,0 +1,261 @@
+
+===Motivation===
+
+Partially Signed Bitcoin Transaction Version 2 as described in BIP370 is not compatible with sending to silent payments as described in BIP352. 
```
This sentence is ambiguous. It makes the reader (or at least me) expect that this BIP introduces version 3. There's two things happening IIUC:

1. The BIP introduces new PSBT fields for Silent Payments
2. Silent Payments by their nature require PSBT version >= 2
